linkerd/proxy/transport/src/listen.rs

```diff
@@ -1,39 +1,111 @@
 use futures::{try_ready, Poll};
+use linkerd2_proxy_core::listen;
 use std::net::SocketAddr;
 use std::time::Duration;
+use tokio::net::TcpStream;
 use tokio::reactor;
+use tracing::trace;
 
-pub struct Listen {
-    local_addr: SocketAddr,
+/// A mockable source for address info, i.e., for tests.
+pub trait OrigDstAddr: Clone {
+    fn orig_dst_addr(&self, socket: &TcpStream) -> Option<SocketAddr>;
+}
+
+#[derive(Clone, Debug)]
+pub struct Bind<O: OrigDstAddr = NoOrigDstAddr> {
+    bind_addr: SocketAddr,
+    keepalive: Option<Duration>,
+    orig_dst_addr: O,
+}
+
+#[derive(Debug)]
+pub struct Listen<O: OrigDstAddr = NoOrigDstAddr> {
```

Comment:
I feel like the naming for this struct is better suited for a trait. How about `Listener`?

Reply:
This is pretty idiomatic in this codebase (we don't have any Connector types, for instance). I'm not opposed to changing the idiom, but don't want to do so ad-hoc.

Reply:
Hm okay good point. No problem